# Post-processing CSV

In [1]:
# https://ipython.org/ipython-doc/3/config/extensions/autoreload.html
%load_ext autoreload
%autoreload 2

In [2]:
from datetime import datetime


import cv2
import numpy as np
import matplotlib.pylab as plt

%matplotlib inline

/usr/local/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')


In [3]:
import sys
sys.path.append("../common/")

In [4]:
from data_utils import get_scalers, LABELS, TRAIN_IMAGE_IDS
from image_utils import get_image_data
from submission_utils import submission_iterator, write_shp_from_polygons, write_shp_from_csv, write_shp_from_mask
from submission_utils import get_scaled_polygons, get_polygons, compute_label_image, get_data_csv, get_image_ids
from submission_utils import rewrite_submission
from visu_utils import plt_st, display_img_1b

In [5]:
for i, l in enumerate(LABELS):
    print i, l

0 None
1 Buildings
2 Misc. Manmade structures
3 Road
4 Track
5 Trees
6 Crops
7 Waterway
8 Standing water
9 Vehicle Large
10 Vehicle Small


Compare submission's predictions raw and post-processed on trained data with ground truth

In [6]:
from np_metrics import score

Get common image ids from submission file and training image ids

In [7]:
def write_shp_pred_pp(image_id, csv_file):
    data_csv = get_data_csv(image_id, csv_file)
    assert data_csv is not None, "No data_csv"
    write_shp_from_csv("../results/shp/%s_new.shp" % image_id, data_csv)
    polygons = get_scaled_polygons(data_csv)
    y_pred_raw = compute_label_image(image_id, polygons)
    y_pred_pp = mask_postprocessing(y_pred_raw, class_pp_func_list)
    write_shp_from_mask("../results/shp/%s_pp.shp" % image_id, image_id, y_pred_pp)

In [8]:
def compare_true_pred_pp(image_id, csv_file, write_shp=True, visualize_labels=()):
    y_true = get_image_data(image_id, 'label')
    data_csv = get_data_csv(image_id, csv_file)
    assert data_csv is not None, "No data_csv"
    if write_shp:
        write_shp_from_csv("../results/shp/%s_new.shp" % image_id, data_csv)
    polygons = get_scaled_polygons(data_csv)    
    y_pred_raw = compute_label_image(image_id, polygons)
    
    h = min(y_true.shape[0], y_pred_raw.shape[0])
    w = min(y_true.shape[1], y_pred_raw.shape[1])
    s1 = score(y_true[:h,:w,:], y_pred_raw[:h,:w,:])    
    print "True vs Pred Raw: ", s1
    
    polygons = shape_postprocessing(polygons, class_shape_pp_func_list)
    y_pred_pp = compute_label_image(image_id, polygons)
    y_pred_pp = mask_postprocessing(y_pred_pp, class_pp_func_list)
    
    h = min(y_true.shape[0], y_pred_pp.shape[0])
    w = min(y_true.shape[1], y_pred_pp.shape[1])
    s2 = score(y_true[:h,:w,:], y_pred_pp[:h,:w,:])
    print "True vs Pred PP: ", s2
    if write_shp:
        write_shp_from_mask("../results/shp/%s_pp.shp" % image_id, image_id, y_pred_pp)
        
        
        
    if len(visualize_labels) > 0:
        ll = len(visualize_labels)
        for i in visualize_labels:
            if i % 2 == 0:
                plt_st(12, 6)
            plt.subplot(131)
            display_img_1b(y_true[:,:,i])
            plt.title("Truth")
            plt.subplot(132)
            display_img_1b(y_pred_raw[:,:,i])
            plt.title("Raw")
            plt.subplot(133)
            display_img_1b(y_pred_pp[:,:,i])
            plt.title("PP")
        
        
        
    return s1, s2

In [10]:
CSV_FILE_NEW = "../results/sub_mix_20170304_2329.csv"
# CSV_FILE_NEW = "../results/submission_2017-03-05-23-26.csv"
CSV_FILE_TRAIN = "../results/sub_model_1_20170304_1826_nan.csv"


In [49]:
from postprocessing_utils import mask_postprocessing
from postprocessing_utils import trees_postprocessing, buildings_postprocessing, roads_postprocessing, standing_water_postprocessing
from postprocessing_utils import shape_postprocessing, roads_shape_postprocessing
from postprocessing_utils import buildings_shape_postprocessing


class_pp_func_list = {
    1: buildings_postprocessing, 
#     3: roads_postprocessing,
    5: trees_postprocessing, 
    8: standing_water_postprocessing, 
}

class_shape_pp_func_list = {
    1: buildings_shape_postprocessing,
    3: roads_shape_postprocessing
}

Check on all training data:

In [13]:
train_ids = TRAIN_IMAGE_IDS

In [14]:
total_s1 = 0
total_s2 = 0
for image_id in train_ids:
    print "--", image_id
    s1, s2 = compare_true_pred_pp(image_id, CSV_FILE_TRAIN, write_shp=False)
    total_s1 += s1
    total_s2 += s2

total_s1 *= 1.0 / len(train_ids)
total_s2 *= 1.0 / len(train_ids)
print "Final score: ", total_s1, total_s2

-- 6040_2_2
True vs Pred Raw:  0.124523
True vs Pred PP:  0.124523
-- 6120_2_2
True vs Pred Raw:  0.301466
True vs Pred PP:  0.303625
-- 6120_2_0
True vs Pred Raw:  0.332018
True vs Pred PP:  0.335627
-- 6090_2_0
True vs Pred Raw:  0.14743
True vs Pred PP:  0.14743
-- 6040_1_3
True vs Pred Raw:  0.137345
True vs Pred PP:  0.137345
-- 6040_1_0
True vs Pred Raw:  0.145944
True vs Pred PP:  0.145954
-- 6100_1_3
True vs Pred Raw:  0.349125
True vs Pred PP:  0.363135
-- 6010_4_2
True vs Pred Raw:  0.131629
True vs Pred PP:  0.131629
-- 6110_4_0
True vs Pred Raw:  0.398946
True vs Pred PP:  0.404301
-- 6140_3_1
True vs Pred Raw:  0.335387
True vs Pred PP:  0.341293
-- 6110_1_2
True vs Pred Raw:  0.233097
True vs Pred PP:  0.237071
-- 6100_2_3
True vs Pred Raw:  0.293448
True vs Pred PP:  0.305344
-- 6150_2_3
True vs Pred Raw:  0.2014
True vs Pred PP:  0.2014
-- 6160_2_1
True vs Pred Raw:  0.139019
True vs Pred PP:  0.139031
-- 6140_1_2
True vs Pred Raw:  0.243752
True vs Pred PP:  0.245974
-

Standing water

-- 6040_2_2
True vs Pred Raw:  0.124523
True vs Pred PP:  0.124523
-- 6120_2_2
True vs Pred Raw:  0.301466
True vs Pred PP:  0.302711
-- 6120_2_0
True vs Pred Raw:  0.332018
True vs Pred PP:  0.332018
-- 6090_2_0
True vs Pred Raw:  0.14743
True vs Pred PP:  0.155803
-- 6040_1_3
True vs Pred Raw:  0.137345
True vs Pred PP:  0.137345
-- 6040_1_0
True vs Pred Raw:  0.145944
True vs Pred PP:  0.145944
-- 6100_1_3
True vs Pred Raw:  0.349125
True vs Pred PP:  0.352157
-- 6010_4_2
True vs Pred Raw:  0.131629
True vs Pred PP:  0.131629
-- 6110_4_0
True vs Pred Raw:  0.398946
True vs Pred PP:  0.399476
-- 6140_3_1
True vs Pred Raw:  0.335387
True vs Pred PP:  0.338185
-- 6110_1_2
True vs Pred Raw:  0.233097
True vs Pred PP:  0.23354
-- 6100_2_3
True vs Pred Raw:  0.293448
True vs Pred PP:  0.295865
-- 6150_2_3
True vs Pred Raw:  0.2014
True vs Pred PP:  0.2014
-- 6160_2_1
True vs Pred Raw:  0.139019
True vs Pred PP:  0.139019
-- 6140_1_2
True vs Pred Raw:  0.243752
True vs Pred PP:  0.243999
-- 6110_3_1
True vs Pred Raw:  0.355562
True vs Pred PP:  0.358455
-- 6010_4_4
True vs Pred Raw:  0.120205
True vs Pred PP:  0.120205
-- 6170_2_4
True vs Pred Raw:  0.220104
True vs Pred PP:  0.219598
-- 6170_4_1
True vs Pred Raw:  0.117173
True vs Pred PP:  0.117173
-- 6170_0_4
True vs Pred Raw:  0.1282
True vs Pred PP:  0.1282
-- 6060_2_3
True vs Pred Raw:  0.183337
True vs Pred PP:  0.183337
-- 6070_2_3
True vs Pred Raw:  0.213583
True vs Pred PP:  0.212168
-- 6010_1_2
True vs Pred Raw:  0.119722
True vs Pred PP:  0.119722
-- 6040_4_4
True vs Pred Raw:  0.121834
True vs Pred PP:  0.121834
-- 6100_2_2
True vs Pred Raw:  0.256762
True vs Pred PP:  0.256735
Final score:  0.214040441513 0.214841640592



Buildings + Roads1 + Trees + Water

-- 6040_2_2
True vs Pred Raw:  0.124523
True vs Pred PP:  0.124417
-- 6120_2_2
True vs Pred Raw:  0.301466
True vs Pred PP:  0.300468
-- 6120_2_0
True vs Pred Raw:  0.332018
True vs Pred PP:  0.333829
-- 6090_2_0
True vs Pred Raw:  0.14743
True vs Pred PP:  0.15652
-- 6040_1_3
True vs Pred Raw:  0.137345
True vs Pred PP:  0.136793
-- 6040_1_0
True vs Pred Raw:  0.145944
True vs Pred PP:  0.143904
-- 6100_1_3
True vs Pred Raw:  0.349125
True vs Pred PP:  0.36002
-- 6010_4_2
True vs Pred Raw:  0.131629
True vs Pred PP:  0.132173
-- 6110_4_0
True vs Pred Raw:  0.398946
True vs Pred PP:  0.406208
-- 6140_3_1
True vs Pred Raw:  0.335387
True vs Pred PP:  0.343092
-- 6110_1_2
True vs Pred Raw:  0.233097
True vs Pred PP:  0.23561
-- 6100_2_3
True vs Pred Raw:  0.293448
True vs Pred PP:  0.297894
-- 6150_2_3
True vs Pred Raw:  0.2014
True vs Pred PP:  0.204052
-- 6160_2_1
True vs Pred Raw:  0.139019
True vs Pred PP:  0.1398
-- 6140_1_2
True vs Pred Raw:  0.243752
True vs Pred PP:  0.246068
-- 6110_3_1
True vs Pred Raw:  0.355562
True vs Pred PP:  0.36527
-- 6010_4_4
True vs Pred Raw:  0.120205
True vs Pred PP:  0.120364
-- 6170_2_4
True vs Pred Raw:  0.220104
True vs Pred PP:  0.218764
-- 6170_4_1
True vs Pred Raw:  0.117173
True vs Pred PP:  0.117373
-- 6170_0_4
True vs Pred Raw:  0.1282
True vs Pred PP:  0.12805
-- 6060_2_3
True vs Pred Raw:  0.183337
True vs Pred PP:  0.185514
-- 6070_2_3
True vs Pred Raw:  0.213583
True vs Pred PP:  0.21341
-- 6010_1_2
True vs Pred Raw:  0.119722
True vs Pred PP:  0.120306
-- 6040_4_4
True vs Pred Raw:  0.121834
True vs Pred PP:  0.120114
-- 6100_2_2
True vs Pred Raw:  0.256762
True vs Pred PP:  0.262434
Final score:  0.214040441513 0.21649792999

Save new submission

In [50]:
submission_file = '../results/submission_' + str(datetime.now().strftime("%Y-%m-%d-%H-%M")) + '.csv'
rewrite_submission(CSV_FILE_NEW, submission_file, class_pp_func_list, class_shape_pp_func_list)

-- 6120_2_4 10 0


Self-intersection at or near point 1331 2425.75


-- 6120_2_3 10 1


Self-intersection at or near point 3195 1987.4666666666667


-- 6120_2_1 10 2
-- 6180_2_4 10 3


Ring Self-intersection at or near point 2934 2924


-- 6180_2_1 10 4


Ring Self-intersection at or near point 2404 2485


-- 6180_2_0 10 5


Ring Self-intersection at or near point 398 2941


-- 6180_2_3 10 6


Ring Self-intersection at or near point 1481 2955


-- 6180_2_2 10 7


Ring Self-intersection at or near point 970 3016


-- 6180_0_3 10 8


Ring Self-intersection at or near point 2033 3085


-- 6180_0_2 10 9


Ring Self-intersection at or near point 981 3118


-- 6180_0_1 10 10


Ring Self-intersection at or near point 2622 2848


-- 6180_0_0 10 11


Ring Self-intersection at or near point 377 2029


-- 6180_0_4 10 12


Ring Self-intersection at or near point 2974 2580


-- 6080_4_4 10 13


Ring Self-intersection at or near point 35 1004


-- 6080_4_2 10 14
-- 6080_4_3 10 15
-- 6080_4_0 10 16


Ring Self-intersection at or near point 2239 2577


-- 6080_4_1 10 17


Ring Self-intersection at or near point 2180 1469


-- 6090_4_1 10 18


Ring Self-intersection at or near point 3206 263


-- 6090_4_0 10 19


Ring Self-intersection at or near point 333 2323


-- 6090_4_3 10 20


Ring Self-intersection at or near point 141 1345


-- 6090_4_2 10 21
-- 6090_4_4 10 22


Ring Self-intersection at or near point 2421 301


-- 6180_4_4 10 23


Ring Self-intersection at or near point 247 2857


-- 6180_4_3 10 24


Ring Self-intersection at or near point 3371 2848


-- 6180_4_2 10 25


Ring Self-intersection at or near point 183 2690


-- 6180_4_1 10 26


Ring Self-intersection at or near point 2955 2889


-- 6180_4_0 10 27


Ring Self-intersection at or near point 1061 2028


-- 6160_3_2 10 28
-- 6160_3_3 10 29
-- 6160_3_0 10 30
-- 6160_3_1 10 31


Ring Self-intersection at or near point 1649 877


-- 6160_3_4 10 32
-- 6080_2_4 10 33
-- 6080_2_0 10 34


Ring Self-intersection at or near point 3089 1573


-- 6080_2_1 10 35


Ring Self-intersection at or near point 1960 3276


-- 6080_2_2 10 36
-- 6080_2_3 10 37
-- 6080_0_2 10 38


Ring Self-intersection at or near point 2162 1716


-- 6080_0_3 10 39


Ring Self-intersection at or near point 2341 2024


-- 6080_0_0 10 40
-- 6080_0_1 10 41
-- 6080_0_4 10 42


Self-intersection at or near point 2892.031007751938 2870.1860465116279


-- 6010_0_4 10 43
-- 6010_0_1 10 44


Ring Self-intersection at or near point 3030 1855


-- 6010_0_0 10 45
-- 6010_0_3 10 46


Ring Self-intersection at or near point 489 2581


-- 6010_0_2 10 47
-- 6010_2_3 10 48


Ring Self-intersection at or near point 2955 2968


-- 6010_2_2 10 49
-- 6010_2_1 10 50
-- 6010_2_0 10 51
-- 6010_2_4 10 52


Self-intersection at or near point 642.10559006211179 2215.9565217391305


-- 6010_4_1 10 53
-- 6170_3_4 10 54


Self-intersection at or near point 33.073619631901842 3056.0797546012268


-- 6010_4_3 10 55


Ring Self-intersection at or near point 3079 2530


-- 6170_3_1 10 56


Ring Self-intersection at or near point 548 3007


-- 6170_3_0 10 57


Ring Self-intersection at or near point 3271 2559


-- 6170_3_3 10 58


Ring Self-intersection at or near point 643 3237


-- 6170_3_2 10 59
-- 6170_1_3 10 60


Ring Self-intersection at or near point 2202 3134


-- 6170_1_2 10 61


Ring Self-intersection at or near point 479 1693


-- 6170_1_1 10 62


Ring Self-intersection at or near point 3077 2846


-- 6170_1_0 10 63


Ring Self-intersection at or near point 1536 1000


-- 6170_1_4 10 64


Ring Self-intersection at or near point 557 3099


-- 6130_4_2 10 65
-- 6130_4_3 10 66
-- 6130_4_0 10 67
-- 6130_4_1 10 68
-- 6130_4_4 10 69


Ring Self-intersection at or near point 2564 633


-- 6150_3_3 10 70
-- 6150_3_2 10 71


Ring Self-intersection at or near point 3320 2799


-- 6150_3_1 10 72


Ring Self-intersection at or near point 2971 137


-- 6150_3_0 10 73


Self-intersection at or near point 168.19999999999999 3258.5999999999999


-- 6150_3_4 10 74
-- 6130_2_0 10 75


Ring Self-intersection at or near point 221 70
Ring Self-intersection at or near point 355 301


-- 6130_2_1 10 76


Ring Self-intersection at or near point 519 2877


-- 6130_2_2 10 77
-- 6130_2_3 10 78
-- 6130_2_4 10 79


Self-intersection at or near point 3023.1116279069765 2775.1069767441859


-- 6130_0_4 10 80
-- 6130_0_2 10 81
-- 6130_0_3 10 82
-- 6130_0_0 10 83
-- 6130_0_1 10 84
-- 6150_1_4 10 85
-- 6150_1_1 10 86


Ring Self-intersection at or near point 1604 2606
Ring Self-intersection at or near point 952 1809


-- 6150_1_0 10 87
-- 6150_1_3 10 88
-- 6150_1_2 10 89


Ring Self-intersection at or near point 1721 1256


-- 6180_3_4 10 90


Ring Self-intersection at or near point 3275 2204


-- 6180_3_0 10 91


Ring Self-intersection at or near point 1459 2826


-- 6180_3_1 10 92


Ring Self-intersection at or near point 968 2609


-- 6180_3_2 10 93


Ring Self-intersection at or near point 1566 3045


-- 6180_3_3 10 94


Ring Self-intersection at or near point 998 2849


-- 6180_1_2 10 95


Ring Self-intersection at or near point 800 542


-- 6180_1_3 10 96


Ring Self-intersection at or near point 1479 3109


-- 6180_1_0 10 97


Ring Self-intersection at or near point 555 2457


-- 6180_1_1 10 98


Ring Self-intersection at or near point 1872 2721


-- 6180_1_4 10 99


Ring Self-intersection at or near point 2578 3328


-- 6010_1_4 10 100


Ring Self-intersection at or near point 3344 1155


-- 6010_1_0 10 101


Ring Self-intersection at or near point 1548 1874


-- 6010_1_1 10 102


Ring Self-intersection at or near point 3133 2171


-- 6010_1_2 10 103
-- 6010_1_3 10 104


Ring Self-intersection at or near point 2501 2900


-- 6010_3_2 10 105


Ring Self-intersection at or near point 2651 3123


-- 6010_3_3 10 106
-- 6010_3_0 10 107


Ring Self-intersection at or near point 2683 3232


-- 6010_3_1 10 108
-- 6010_3_4 10 109


Ring Self-intersection at or near point 341 2046
Self-intersection at or near point 2185.5288461538462 1021.0961538461538


-- 6020_0_4 10 110
-- 6020_0_0 10 111
-- 6020_0_1 10 112
-- 6020_0_2 10 113
-- 6020_0_3 10 114
-- 6020_2_2 10 115
-- 6020_2_3 10 116
-- 6020_2_0 10 117


Ring Self-intersection at or near point 2079 2951


-- 6020_2_1 10 118
-- 6020_2_4 10 119


Ring Self-intersection at or near point 2460 2035


-- 6020_4_0 10 120
-- 6020_4_1 10 121


Ring Self-intersection at or near point 2835 2868


-- 6020_4_2 10 122
-- 6020_4_3 10 123


Ring Self-intersection at or near point 2385 1563


-- 6020_4_4 10 124


Ring Self-intersection at or near point 2978 3337


-- 6110_0_1 10 125
-- 6150_0_4 10 126
-- 6150_0_0 10 127
-- 6150_0_1 10 128


Ring Self-intersection at or near point 1158 3324


-- 6150_0_2 10 129
-- 6150_0_3 10 130
-- 6100_2_1 10 131
-- 6100_2_0 10 132
-- 6150_2_2 10 133


Ring Self-intersection at or near point 2469 2386


-- 6150_2_0 10 134


Ring Self-intersection at or near point 3221 2085


-- 6150_2_1 10 135
-- 6150_2_4 10 136
-- 6170_4_2 10 137


Ring Self-intersection at or near point 673 2340


-- 6050_3_1 10 138
-- 6170_4_3 10 139


Ring Self-intersection at or near point 2994 3094


-- 6150_4_0 10 140
-- 6150_4_1 10 141


Ring Self-intersection at or near point 2086 2494
Self-intersection at or near point 3218.0202020202019 3106.8989898989898


-- 6150_4_2 10 142
-- 6150_4_3 10 143
-- 6150_4_4 10 144
-- 6170_4_0 10 145


Ring Self-intersection at or near point 2309 3060


-- 6070_3_4 10 146


Ring Self-intersection at or near point 2693 2323


-- 6070_3_0 10 147
-- 6070_3_1 10 148
-- 6070_3_2 10 149


Ring Self-intersection at or near point 2086 1749


-- 6070_3_3 10 150
-- 6070_1_2 10 151


Ring Self-intersection at or near point 2737 2311


-- 6070_1_3 10 152


Ring Self-intersection at or near point 749 2450


-- 6070_1_0 10 153
-- 6070_1_1 10 154


Ring Self-intersection at or near point 650 3112


-- 6070_1_4 10 155


Ring Self-intersection at or near point 3138 2523


-- 6120_0_3 10 156


Self-intersection at or near point 2379.0055248618783 25.966850828729282


-- 6110_1_1 10 157
-- 6110_1_0 10 158
-- 6110_1_3 10 159
-- 6110_1_4 10 160
-- 6110_3_4 10 161
-- 6110_3_3 10 162


Ring Self-intersection at or near point 2358 801


-- 6110_3_2 10 163
-- 6110_3_0 10 164


Self-intersection at or near point 2033.5604395604396 1757.4945054945056


-- 6060_2_1 10 165
-- 6140_3_3 10 166
-- 6020_1_4 10 167


Self-intersection at or near point 3116.0972222222222 1686.4166666666667
Ring Self-intersection at or near point 2714 2512


-- 6020_1_1 10 168
-- 6020_1_0 10 169


Ring Self-intersection at or near point 778 3157


-- 6020_1_3 10 170


Ring Self-intersection at or near point 171 690


-- 6020_1_2 10 171
-- 6020_3_3 10 172
-- 6020_3_2 10 173


Self-intersection at or near point 3310.344827586207 1894.344827586207


-- 6020_3_1 10 174
-- 6020_3_0 10 175


Ring Self-intersection at or near point 1458 3321


-- 6020_3_4 10 176
-- 6050_2_4 10 177
-- 6050_2_3 10 178
-- 6050_2_2 10 179
-- 6050_2_1 10 180
-- 6050_2_0 10 181
-- 6100_3_4 10 182
-- 6100_3_0 10 183
-- 6100_3_1 10 184
-- 6100_3_2 10 185
-- 6100_3_3 10 186


Self-intersection at or near point 2989.1764705882351 3.2549019607843137
Self-intersection at or near point 2593.4285714285716 303.52380952380952


-- 6060_1_1 10 187
-- 6060_1_0 10 188
-- 6060_1_3 10 189


Self-intersection at or near point 2986.3617021276596 1244.4468085106382


-- 6060_1_2 10 190


Self-intersection at or near point 1358 1683.4285714285713


-- 6060_1_4 10 191
-- 6060_3_4 10 192


Self-intersection at or near point 1163.159090909091 1174.8181818181818


-- 6060_3_3 10 193


Self-intersection at or near point 559 3332


-- 6060_3_2 10 194
-- 6060_3_1 10 195


Self-intersection at or near point 3399.2655367231637 628.62146892655369


-- 6060_3_0 10 196
-- 6070_2_4 10 197


Ring Self-intersection at or near point 449 1356


-- 6070_2_1 10 198
-- 6070_2_0 10 199
-- 6070_2_3 10 200


Self-intersection at or near point 1791 3061
Ring Self-intersection at or near point 942 2260


-- 6070_2_2 10 201


Ring Self-intersection at or near point 2621 2731


-- 6040_2_0 10 202


Ring Self-intersection at or near point 2658 670


-- 6040_2_1 10 203


Ring Self-intersection at or near point 1643 2761


-- 6040_2_3 10 204


Ring Self-intersection at or near point 2416 1482


-- 6040_2_4 10 205


Ring Self-intersection at or near point 1778 56


-- 6070_0_3 10 206


Ring Self-intersection at or near point 2716 3224


-- 6070_0_2 10 207
-- 6070_0_1 10 208
-- 6070_0_0 10 209


Self-intersection at or near point 3220 3124


-- 6070_0_4 10 210
-- 6100_4_3 10 211
-- 6040_0_4 10 212


Ring Self-intersection at or near point 1148 3067


-- 6040_0_2 10 213
-- 6040_0_3 10 214


Ring Self-intersection at or near point 2973 3207


-- 6040_0_0 10 215


Ring Self-intersection at or near point 1288 1463


-- 6040_0_1 10 216


Ring Self-intersection at or near point 3021 2143


-- 6050_0_1 10 217
-- 6050_0_0 10 218


Ring Self-intersection at or near point 1577 710


-- 6050_0_3 10 219
-- 6050_0_2 10 220
-- 6050_0_4 10 221
-- 6070_4_4 10 222
-- 6070_4_3 10 223


Ring Self-intersection at or near point 787 3155


-- 6070_4_2 10 224


Ring Self-intersection at or near point 1570 2329
Self-intersection at or near point 768 972


-- 6070_4_1 10 225
-- 6070_4_0 10 226


Self-intersection at or near point 3121 2140


-- 6140_3_0 10 227
-- 6140_3_2 10 228
-- 6100_4_4 10 229
-- 6140_3_4 10 230
-- 6100_4_2 10 231
-- 6100_4_1 10 232
-- 6100_4_0 10 233
-- 6140_0_1 10 234
-- 6140_1_4 10 235
-- 6140_1_3 10 236
-- 6140_1_0 10 237


Self-intersection at or near point 3249 833


-- 6140_1_1 10 238
-- 6040_4_2 10 239


Ring Self-intersection at or near point 2768 1198


-- 6040_4_3 10 240


Ring Self-intersection at or near point 2540 2077


-- 6040_4_0 10 241
-- 6040_4_1 10 242


Ring Self-intersection at or near point 2474 3043


-- 6040_4_4 10 243


Ring Self-intersection at or near point 547 3120


-- 6010_4_0 10 244
-- 6110_0_0 10 245
-- 6050_4_4 10 246
-- 6110_0_2 10 247
-- 6110_0_3 10 248
-- 6110_0_4 10 249


Ring Self-intersection at or near point 1113 2901


-- 6050_4_0 10 250
-- 6050_4_3 10 251


Self-intersection at or near point 2864.6756756756758 414.94594594594594


-- 6050_4_2 10 252
-- 6030_3_0 10 253


Ring Self-intersection at or near point 781 1667
Ring Self-intersection at or near point 834 1859


-- 6030_3_1 10 254
-- 6030_3_2 10 255


Ring Self-intersection at or near point 1699 2


-- 6030_3_3 10 256


Self-intersection at or near point 3217 340


-- 6030_3_4 10 257


Self-intersection at or near point 1938.9230769230769 629.61538461538464
Ring Self-intersection at or near point 1791 1058


-- 6110_2_4 10 258
-- 6110_2_2 10 259
-- 6110_2_3 10 260
-- 6110_2_0 10 261


Ring Self-intersection at or near point 1517 2689


-- 6110_2_1 10 262
-- 6120_4_1 10 263


Ring Self-intersection at or near point 39 1872


-- 6120_4_0 10 264


Self-intersection at or near point 3268.7127071823206 1381.9779005524863


-- 6120_4_3 10 265
-- 6120_4_2 10 266
-- 6110_4_4 10 267
-- 6110_4_1 10 268


Self-intersection at or near point 72.881720430107521 1000.989247311828


-- 6110_4_2 10 269
-- 6110_4_3 10 270


Ring Self-intersection at or near point 2589 2017


-- 6060_4_4 10 271
-- 6060_4_0 10 272


Ring Self-intersection at or near point 656 2982


-- 6060_4_1 10 273


Ring Self-intersection at or near point 2041 1315


-- 6060_4_2 10 274


Ring Self-intersection at or near point 2394 3298


-- 6060_4_3 10 275


Ring Self-intersection at or near point 2878 1981
Self-intersection at or near point 383.89743589743591 2890.1794871794873


-- 6100_2_4 10 276
-- 6050_3_4 10 277
-- 6050_3_2 10 278
-- 6050_3_3 10 279
-- 6050_3_0 10 280
-- 6100_2_2 10 281


Self-intersection at or near point 2397.04 168.80000000000001


-- 6100_0_3 10 282


Ring Self-intersection at or near point 2915 2891
Ring Self-intersection at or near point 2545 1360
Ring Self-intersection at or near point 3308 3223


-- 6100_0_2 10 283


Ring Self-intersection at or near point 2199 2481


-- 6100_0_1 10 284
-- 6100_0_0 10 285
-- 6100_0_4 10 286
-- 6060_0_0 10 287


Self-intersection at or near point 859 274


-- 6060_0_1 10 288


Ring Self-intersection at or near point 3298 1942


-- 6060_0_2 10 289


Ring Self-intersection at or near point 1097 2590
Ring Self-intersection at or near point 974 1860


-- 6060_0_3 10 290
-- 6060_0_4 10 291


Ring Self-intersection at or near point 995 654


-- 6160_1_0 10 292


Ring Self-intersection at or near point 3217 3164


-- 6060_2_4 10 293


Ring Self-intersection at or near point 1411 487


-- 6060_2_2 10 294
-- 6060_2_0 10 295
-- 6160_1_1 10 296


Ring Self-intersection at or near point 2858 3199


-- 6040_3_1 10 297


Ring Self-intersection at or near point 2196 2015


-- 6040_3_0 10 298


Ring Self-intersection at or near point 1821 1178


-- 6040_3_3 10 299


Ring Self-intersection at or near point 1775 2772


-- 6040_3_2 10 300


Ring Self-intersection at or near point 500 1945


-- 6040_3_4 10 301


Ring Self-intersection at or near point 1726 1901


-- 6090_0_4 10 302
-- 6160_1_3 10 303


Ring Self-intersection at or near point 1119 155


-- 6090_0_1 10 304


Ring Self-intersection at or near point 1821 1447


-- 6090_0_0 10 305


Ring Self-intersection at or near point 2854 1057


-- 6090_0_3 10 306


Ring Self-intersection at or near point 1467 3102


-- 6090_0_2 10 307


Ring Self-intersection at or near point 2475 504


-- 6090_2_3 10 308
-- 6090_2_2 10 309
-- 6090_2_1 10 310


Ring Self-intersection at or near point 2160 3194


-- 6040_1_4 10 311


Ring Self-intersection at or near point 2369 2286


-- 6040_1_2 10 312


Self-intersection at or near point 131.05405405405406 2567.1891891891892


-- 6040_1_1 10 313
-- 6090_2_4 10 314


Ring Self-intersection at or near point 3319 1397


-- 6030_0_4 10 315


Ring Self-intersection at or near point 1759 2343


-- 6030_0_3 10 316


Self-intersection at or near point 2737 2587.25
Ring Self-intersection at or near point 518 2633


-- 6030_0_2 10 317
-- 6030_0_1 10 318
-- 6030_0_0 10 319


Ring Self-intersection at or near point 2345 2114


-- 6050_1_0 10 320
-- 6050_1_1 10 321
-- 6050_1_2 10 322
-- 6050_1_3 10 323
-- 6050_1_4 10 324
-- 6120_1_0 10 325


Ring Self-intersection at or near point 1885 47


-- 6120_1_1 10 326
-- 6120_1_2 10 327
-- 6120_1_3 10 328
-- 6120_1_4 10 329
-- 6120_0_4 10 330
-- 6160_1_4 10 331


Ring Self-intersection at or near point 878 2826
Self-intersection at or near point 982 1493


-- 6130_3_0 10 332
-- 6140_0_4 10 333
-- 6140_0_3 10 334
-- 6140_0_2 10 335
-- 6160_1_2 10 336


Ring Self-intersection at or near point 2105 2857


-- 6140_0_0 10 337
-- 6120_3_4 10 338


Ring Self-intersection at or near point 1795 3030
Self-intersection at or near point 2518 1822.5


-- 6120_3_2 10 339


Ring Self-intersection at or near point 2872 2251
Self-intersection at or near point 891.22641509433959 2429.8301886792451


-- 6120_3_3 10 340


Ring Self-intersection at or near point 880 2463


-- 6120_3_0 10 341


Ring Self-intersection at or near point 2873 2809


-- 6120_3_1 10 342


Self-intersection at or near point 1021.4084507042254 1378.0985915492959


-- 6100_1_2 10 343


Self-intersection at or near point 2754.1100000000001 2461.4499999999998
Ring Self-intersection at or near point 1359 2307


-- 6100_1_0 10 344


Ring Self-intersection at or near point 1747 2578


-- 6100_1_1 10 345
-- 6100_1_4 10 346
-- 6140_4_3 10 347
-- 6140_4_2 10 348
-- 6140_4_1 10 349


Self-intersection at or near point 1142.6666666666667 179.16666666666666
Self-intersection at or near point 3310.7228915662649 1497.301204819277


-- 6140_4_0 10 350
-- 6140_4_4 10 351
-- 6160_0_4 10 352


Ring Self-intersection at or near point 1524 3229


-- 6140_2_1 10 353
-- 6140_2_0 10 354


Self-intersection at or near point 1845.0714285714287 993.21428571428567


-- 6140_2_3 10 355


Ring Self-intersection at or near point 1191 530


-- 6140_2_2 10 356
-- 6140_2_4 10 357


Ring Self-intersection at or near point 2615 695


-- 6030_4_3 10 358
-- 6030_4_2 10 359


Self-intersection at or near point 466.969696969697 1306.6969696969697
Ring Self-intersection at or near point 2832 825


-- 6030_4_1 10 360


Self-intersection at or near point 2741 1622


-- 6030_4_0 10 361


Self-intersection at or near point 2380 3211


-- 6030_4_4 10 362
-- 6130_3_4 10 363


Ring Self-intersection at or near point 2753 2220


-- 6030_2_1 10 364
-- 6030_2_0 10 365


Ring Self-intersection at or near point 2064 1540


-- 6030_2_3 10 366
-- 6030_2_2 10 367
-- 6030_2_4 10 368


Self-intersection at or near point 3266 699


-- 6090_3_2 10 369
-- 6090_3_3 10 370
-- 6090_3_0 10 371


Ring Self-intersection at or near point 852 2550


-- 6090_3_1 10 372


Ring Self-intersection at or near point 2524 3261


-- 6090_3_4 10 373


Ring Self-intersection at or near point 1368 1777


-- 6080_3_4 10 374
-- 6170_4_4 10 375


Ring Self-intersection at or near point 180 2339


-- 6080_3_1 10 376
-- 6080_3_0 10 377
-- 6080_3_3 10 378


Ring Self-intersection at or near point 2145 3337


-- 6080_3_2 10 379
-- 6160_4_1 10 380


Ring Self-intersection at or near point 3079 761


-- 6160_4_0 10 381


Ring Self-intersection at or near point 1371 3298


-- 6160_4_3 10 382


Ring Self-intersection at or near point 489 2711


-- 6160_4_2 10 383


Ring Self-intersection at or near point 28 640


-- 6160_4_4 10 384
-- 6080_1_3 10 385
-- 6080_1_2 10 386


Ring Self-intersection at or near point 2413 1419


-- 6080_1_1 10 387
-- 6080_1_0 10 388
-- 6080_1_4 10 389
-- 6170_2_0 10 390


Ring Self-intersection at or near point 1258 3017


-- 6170_2_1 10 391
-- 6170_2_2 10 392


Ring Self-intersection at or near point 1003 3073


-- 6170_2_3 10 393
-- 6170_0_2 10 394


Ring Self-intersection at or near point 1741 2026


-- 6170_0_3 10 395


Ring Self-intersection at or near point 2760 1300


-- 6170_0_0 10 396


Ring Self-intersection at or near point 3207 2474


-- 6170_0_1 10 397


Ring Self-intersection at or near point 61 1861


-- 6090_1_4 10 398


Ring Self-intersection at or near point 167 2470


-- 6120_4_4 10 399
-- 6090_1_0 10 400


Ring Self-intersection at or near point 339 2442


-- 6090_1_1 10 401


Ring Self-intersection at or near point 2043 2952


-- 6090_1_2 10 402
-- 6090_1_3 10 403
-- 6120_0_1 10 404


Ring Self-intersection at or near point 1536 3027


-- 6120_0_0 10 405


Ring Self-intersection at or near point 1553 1540


-- 6030_1_4 10 406
-- 6120_0_2 10 407
-- 6030_1_2 10 408


Ring Self-intersection at or near point 32 1823


-- 6030_1_3 10 409
-- 6030_1_0 10 410
-- 6030_1_1 10 411
-- 6130_3_1 10 412
-- 6050_4_1 10 413
-- 6130_3_3 10 414


Ring Self-intersection at or near point 1083 879


-- 6130_3_2 10 415


Ring Self-intersection at or near point 2563 62


-- 6160_0_1 10 416


Ring Self-intersection at or near point 2133 3324


-- 6160_0_0 10 417


Ring Self-intersection at or near point 3231 2503


-- 6160_0_3 10 418
-- 6160_0_2 10 419


Ring Self-intersection at or near point 431 1403


-- 6160_2_3 10 420
-- 6160_2_2 10 421


Ring Self-intersection at or near point 486 965


-- 6160_2_0 10 422
-- 6160_2_4 10 423


Ring Self-intersection at or near point 970 2896


-- 6130_1_4 10 424
-- 6130_1_3 10 425
-- 6130_1_2 10 426


Ring Self-intersection at or near point 3338 2672


-- 6130_1_1 10 427
-- 6130_1_0 10 428


In [14]:
import pandas as pd

In [15]:
df1 = pd.read_csv(CSV_FILE_NEW)
df2 = pd.read_csv(submission_file)

In [16]:
assert (df1['ImageId'] == df2['ImageId']).all(), "WTF"
assert (df1['ClassType'] == df2['ClassType']).all(), "WTF"

In [17]:
# assert (df1['MultipolygonWKT'] == df2['MultipolygonWKT']).all(), "WTF"

Visual data analysis

In [19]:
import gdal
from visu_utils import scale_percentile

def generate_complete_image(image_id):
    filename = os.path.join("..", "results", "visu", '%s.tif' % image_id)
    if os.path.exists(filename):
        print "Image is already created. Remove it"
        os.remove(filename)        

    tile_size = (1024, 1024)
    n, m = 5, 5
    complete_image = np.zeros((5*tile_size[0], 5*tile_size[1]), dtype=np.uint16)
    counter = 0
    for i in range(n):
        ys = i*tile_size[1]
        ye = ys + tile_size[1]
        for j in range(m):
            xs = j*tile_size[0]
            xe = xs + tile_size[0]
            _id = image_id + "_%i_%i" % (i, j)
            img = get_image_data(_id, 'pan')
            img = cv2.resize(img, dsize=tile_size)    
            complete_image[ys:ye, xs:xe] = img

    complete_image_8u = scale_percentile(complete_image, 0.1, 99.9)             
    # Initialize output file:
    driver = gdal.GetDriverByName('GTiff')
    dataset = driver.Create(filename, complete_image_8u.shape[1], complete_image_8u.shape[0], 1, gdal.GDT_Byte)
    assert dataset is not None, "Failed to create a file"

    band = dataset.GetRasterBand(1)
    band.WriteArray(complete_image_8u, xoff=0, yoff=0)

    dataset = None

In [25]:
import os
from submission_utils import get_data_csv
from shapely.wkt import loads
from shapely.affinity import scale, translate

def get_transformed_polygons(data_csv, x, y, w, h):
    out = {}
    image_id = data_csv[0][0]
    xs, ys = get_scalers(image_id, h, w)
    for poly_info in data_csv:
        if "MULTIPOLYGON" not in poly_info[2][:20]:
            continue
        polygons = loads(poly_info[2])
        new_polygons = scale(polygons, xfact=xs, yfact=ys, origin=(0, 0, 0))        
        new_polygons = translate(new_polygons, xoff=x, yoff=y)
        out[int(poly_info[1])] = new_polygons
    return out

from collections import defaultdict

def generate_complete_shp_from_csv(image_id, csv_filename, suffix=''):
    
    filename = os.path.join("..", "results", "visu", '%s_%s.shp' % (image_id, suffix))
    if os.path.exists(filename):
        print "Image is already created. Remove it"
        os.remove(filename)        
        
    tile_size = (1024, 1024)
    n, m = 5, 5
    all_polygons = defaultdict(list)
    counter = 0
    for i in range(n):
        y = i*tile_size[1]
        for j in range(m):
            x = j*tile_size[0]
            _id = image_id + "_%i_%i" % (i, j)
            print "--", _id
            data_csv = get_data_csv(_id, csv_filename)        
            if data_csv is None:
                continue
            polygons = get_transformed_polygons(data_csv, x, y, tile_size[0], tile_size[1])
            for k in polygons:
                all_polygons[k].append(polygons[k])
    write_shp_from_polygons(filename, image_id, all_polygons)


In [54]:
CSV_FILE_LAST = submission_file

In [22]:
# get_image_ids(CSV_FILE_LAST)

In [61]:
image_id = "6110"
# image_id = "6120"
# image_id = "6070"
# image_id = "6080"
# image_id = "6050"

# image_id = "6140"

In [46]:
generate_complete_image(image_id)

In [62]:
generate_complete_shp_from_csv(image_id, CSV_FILE_LAST, CSV_FILE_LAST[-20:-4])

-- 6110_0_0
-- 6110_0_1
-- 6110_0_2
-- 6110_0_3
-- 6110_0_4
-- 6110_1_0
-- 6110_1_1
-- 6110_1_2
-- 6110_1_3
-- 6110_1_4
-- 6110_2_0
-- 6110_2_1
-- 6110_2_2
-- 6110_2_3
-- 6110_2_4
-- 6110_3_0
-- 6110_3_1
-- 6110_3_2
-- 6110_3_3
-- 6110_3_4
-- 6110_4_0
-- 6110_4_1
-- 6110_4_2
-- 6110_4_3
-- 6110_4_4
Written succesfully file :  ../results/visu/6110_2017-03-07-02-17.shp
